sw/device/tests/dif/dif_clkmgr_unittest.cc

```diff
@@ -0,0 +1,293 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_clkmgr.h"
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/testing/mock_mmio.h"
+
+#include "clkmgr_regs.h"
+
+namespace dif_clkmgr_unittest {
+namespace {
+using mock_mmio::MmioTest;
+using mock_mmio::MockDevice;
+using testing::Test;
+
+class ClkMgrTest : public Test, public MmioTest {
+ protected:
+  const dif_clkmgr_gateable_clock_t default_last_gateable_clock = 4;
+  const dif_clkmgr_hintable_clock_t default_last_hintable_clock = 6;
+  const dif_clkmgr_params_t params_ = {
+      .base_addr = dev().region(),
+      .last_gateable_clock = default_last_gateable_clock,
+      .last_hintable_clock = default_last_hintable_clock};
+};
+
+class InitTest : public ClkMgrTest {};
+
+TEST_F(InitTest, Error) {
+  // null handle
+  EXPECT_EQ(dif_clkmgr_init(params_, nullptr), kDifClkmgrBadArg);
+
+  // out-of-bounds last gateable clock
+  {
+    dif_clkmgr_t clkmgr;
+    dif_clkmgr_params_t bad_gateable = params_;
+    bad_gateable.last_gateable_clock = 32;
+    EXPECT_EQ(dif_clkmgr_init(bad_gateable, &clkmgr), kDifClkmgrBadArg);
+  }
+  {
+    dif_clkmgr_t clkmgr;
+    dif_clkmgr_params_t bad_gateable = params_;
+    bad_gateable.last_gateable_clock = ~0;
+    EXPECT_EQ(dif_clkmgr_init(bad_gateable, &clkmgr), kDifClkmgrBadArg);
+  }
+
+  // out-of-bounds last hintable clock
+  {
+    dif_clkmgr_t clkmgr;
+    dif_clkmgr_params_t bad_hintable = params_;
+    bad_hintable.last_hintable_clock = 32;
+    EXPECT_EQ(dif_clkmgr_init(bad_hintable, &clkmgr), kDifClkmgrBadArg);
+  }
+  {
+    dif_clkmgr_t clkmgr;
+    dif_clkmgr_params_t bad_hintable = params_;
+    bad_hintable.last_hintable_clock = ~0;
+    EXPECT_EQ(dif_clkmgr_init(bad_hintable, &clkmgr), kDifClkmgrBadArg);
+  }
+}
+
+class GateableClockTest : public ClkMgrTest {};
+
+TEST_F(GateableClockTest, Set) {
+  dif_clkmgr_t clkmgr;
+  EXPECT_EQ(dif_clkmgr_init(params_, &clkmgr), kDifClkmgrOk);
```

Comment:
You probably could have `dif_clkmgr_t clkmgr` as one of the top class parameter, that would avoid calling `init` for every unit test.